### PR TITLE
Fix 148 join or create

### PR DESCRIPTION
**Pull request by mitchLucid**
Removed the implied expectation that users would provide an `onAddRoom` callback function to the ColyseusClient. The client itself now performs the `.Connect()` call on the room, at the same time where it previously was attempting to invoke the `onAddRoom` callback if it had been provided. Also restructured the `JoinOrCreate` function to match the `Join` and `Create` functions and added a way to trigger it from within the example Lobby

Fixes [Issue 148](https://github.com/colyseus/colyseus-unity3d/issues/148)

### CODE REVIEW

File: Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs [hunk 1]
```diff
@@ -265,7 +265,9 @@ void OnJoin()
 
             onAddRoom?.Invoke(room);
 
-            return await tcs.Task;
+            await room.Connect();
+
+			return await tcs.Task;
```
Comment (endel) [inline Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs:270]: nit: indentation

File: Assets/Colyseus/Runtime/Scripts/Managers/ColyseusManager.cs [hunk 2]
```diff
@@ -138,10 +138,10 @@ public ColyseusClient CreateClient(string endpoint)
...
 
-        /// <summary>
-        /// Connect to the Colyseus server.
-        /// </summary>
-        protected virtual void ConnectToServer()
+		/// <summary>
+		/// Create a new <see cref="ColyseusClient"/> along with any other client initialization you may need to perform
+		/// </summary>
+		public virtual void InitializeClient()
         {
```
Comment (endel) [inline Assets/Colyseus/Runtime/Scripts/Managers/ColyseusManager.cs:145]: nit: indentation

File: Assets/Colyseus/Runtime/Example~/Scripts/Lobby/UI/LobbyController.cs [hunk 3]
```diff
@@ -80,9 +80,18 @@ public void CreateRoom()
...
+        {
+	        LoadGallery(() => { ExampleManager.Instance.CreateNewRoom(desiredRoomName); });
+        }
     }
 
+    public void JoinOrCreateRoom()
+    {
+	    connectingCover.SetActive(true);
+		LoadGallery(() => { ExampleManager.Instance.JoinOrCreateRoom(); });
+	}
```
Comment (endel) [inline Assets/Colyseus/Runtime/Example~/Scripts/Lobby/UI/LobbyController.cs:93]: nit: indentation

File: Assets/Colyseus/Runtime/Example~/Scripts/ExampleRoomController.cs [hunk 4]
```diff
@@ -314,29 +304,32 @@ public async Task AddRoom(IColyseusRoom roomToAdd)
...
-            LSLog.LogImportant($"Joined / Created Room: {_room.Id}");
-            _lastRoomId = _room.Id;
-            RegisterRoomHandlers();
         }
-        catch (Exception e)
+        catch (Exception ex)
         {
-            LSLog.LogError($"Room Controller Error - {e.Message + e.StackTrace}");
+            LSLog.LogError($"Room Controller Error - {ex.Message + ex.StackTrace}");
+			onComplete?.Invoke(false);
```
Comment (endel) [inline Assets/Colyseus/Runtime/Example~/Scripts/ExampleRoomController.cs:324]: nit: indentation

File: Assets/Colyseus/Runtime/Example~/Scripts/ExampleRoomController.cs [hunk 5]
```diff
@@ -314,29 +304,32 @@ public async Task AddRoom(IColyseusRoom roomToAdd)
     ///     Join an existing room or create a new one using <see cref="roomName" /> with no options.
     ///     <para>Locked or private rooms are ignored.</para>
     /// </summary>
-    public async void JoinOrCreateRoom()
+    public async Task JoinOrCreateRoom(Action<bool> onComplete = null)
     {
-        try
+	    LSLog.LogImportant($"Join Or Create Room - Name = {roomName}.... ");
+		try
```
Comment (endel) [inline Assets/Colyseus/Runtime/Example~/Scripts/ExampleRoomController.cs:310]: nit: indentation

File: Assets/Colyseus/Runtime/Example~/Scripts/ExampleManager.cs [hunk 6]
```diff
@@ -158,6 +147,11 @@ public async void CreateNewRoom(string roomID)
         await _roomController.CreateSpecificRoom(client, _roomController.roomName, roomID);
     }
 
+    public async void JoinOrCreateRoom()
+    {
+	    await _roomController.JoinOrCreateRoom();
```
Comment (endel) [inline Assets/Colyseus/Runtime/Example~/Scripts/ExampleManager.cs:152]: nit: indentation (2 spaces, everywhere else seems to be 4 spaces)

**Comment by tonygiang** [comment]
Hey, I just pulled this branch to test out, it's still not entirely fixed. With this branch, the server has received the Join signal and proceeded with the client joining, but the `JoinOrCreate` task still runs indefinitely with the status `WaitingForActivation`.

**Pull request merged**